### PR TITLE
Add subdivisions in ticksscalebar right segments

### PR DESCRIPTION
**Pull request by agiudiceandrea**
## Description

Adds the ability to create subdivisions for segments included in the right part of the ticksscalebar.

Line Ticks Up
image

Line Ticks Middle
image

Line Ticks Down
image

Scalebar item properties - Segments groupbox
image

I have not changed the behaviour related to the segments on the left side of the scalebar.

### CODE REVIEW

File: src/core/layout/qgslayoutitemscalebar.h [hunk 1]
```diff
@@ -81,6 +81,18 @@ class CORE_EXPORT QgsLayoutItemScaleBar: public QgsLayoutItem
      * \see setNumberOfSegments()
      */
     void setNumberOfSegmentsLeft( int segments );
+    
+    /**
+     * Returns the number of subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see setNumberOfSubdivisions()
+     */
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutitemscalebar.h:88]: \since QGIS 3.14

File: src/core/layout/qgslayoutitemscalebar.h [hunk 2]
```diff
@@ -81,6 +81,18 @@ class CORE_EXPORT QgsLayoutItemScaleBar: public QgsLayoutItem
...
+    int numberOfSubdivisions()  const { return mSettings.numberOfSubdivisions(); }
+
+    /**
+     * Sets the number of \a subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see numberOfSubdivisions()
+     */
```
Comment (nyalldawson) [inline src/core/layout/qgslayoutitemscalebar.h:94]: \since QGIS 3.14

File: src/core/scalebar/qgsscalebarsettings.h [hunk 3]
```diff
@@ -120,6 +120,18 @@ class CORE_EXPORT QgsScaleBarSettings
      */
     void setNumberOfSegmentsLeft( int segments ) { mNumSegmentsLeft = segments; }
 
+    /**
+     * Returns the number of subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see setNumberOfSubdivisions()
+     */
```
Comment (nyalldawson) [inline src/core/scalebar/qgsscalebarsettings.h:126]: \since QGIS 3.14

File: src/core/scalebar/qgsscalebarsettings.h [hunk 4]
```diff
@@ -120,6 +120,18 @@ class CORE_EXPORT QgsScaleBarSettings
...
+    int numberOfSubdivisions() const { return mNumSubdivisions; }
+
+    /**
+     * Sets the number of \a subdivisions for segments included in the right part of the scalebar (only used for some scalebar types).
+     * \see numberOfSubdivisions()
+     */
```
Comment (nyalldawson) [inline src/core/scalebar/qgsscalebarsettings.h:132]: \since QGIS 3.14

File: src/core/scalebar/qgsticksscalebarrenderer.cpp [hunk 5]
```diff
@@ -114,13 +115,44 @@ void QgsTicksScaleBarRenderer::draw( QgsRenderContext &context, const QgsScaleBa
...
+    switch ( mTickPosition )
+    {
+      case TicksDown:
+        verticalPos = barTopPosition;
+        subTickPositionsY << barTopPosition;
+        subTickPositionsY << middlePosition;
+        break;
+      case TicksMiddle:
+        verticalPos = middlePosition;
+        subTickPositionsY << (barTopPosition + middlePosition ) / 2;
```
Comment (nyalldawson) [inline src/core/scalebar/qgsticksscalebarrenderer.cpp:137]: I think ideally we'd expose the subdivision height to user control

**Review by nyalldawson — COMMENTED** [review]
Code is looking good! Got some screenshots to share?

File: src/core/scalebar/qgsticksscalebarrenderer.cpp (hunk 5 above)
Comment (agiudiceandrea) [reply src/core/scalebar/qgsticksscalebarrenderer.cpp]: Hi @nyalldawson, thanks for the revision of this preliminary work.
What do you think about letting the user choose the subdivisions tick height <= the segments tick height?

Comment (nyalldawson) [reply src/core/scalebar/qgsticksscalebarrenderer.cpp]: I think it's a good idea. We could either expose it as an absolute mm size or a percentage (of tick height) size...

Comment (agiudiceandrea) [reply src/core/scalebar/qgsticksscalebarrenderer.cpp]: Subdivisions tick height control added.

**Comment by nyalldawson** [comment]
Nice! Looking good to me. Can you add a unit test and then we can merge....

**Review by nyalldawson — APPROVED** [review]

**Pull request merged**